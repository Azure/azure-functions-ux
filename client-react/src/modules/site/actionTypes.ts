export const AREA_STRING = '@site';
export const UPDATE_RESOURCE_ID = `@site/UPDATE_RESOURCE_ID`;
export const SITE_FETCH_REQUEST = `@site/FETCH_REQUEST`;
export const SITE_FETCH_SUCCESS = `@site/FETCH_SUCCESS`;
export const SITE_FETCH_FAILURE = `@site/FETCH_FAILURE`;
export const SITE_UPDATE_REQUEST = `@site/UPDATE_REQUEST`;
export const SITE_UPDATE_SUCCESS = `@site/UPDATE_SUCCESS`;
export const SITE_UPDATE_FAILURE = `@site/UPDATE_FAILURE`;
