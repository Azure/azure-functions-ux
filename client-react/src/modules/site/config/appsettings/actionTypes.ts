export const AREA_STRING = '@appsettings';
export const APP_SETTINGS_FETCH_REQUEST = '@appsettings/FETCH_REQUEST';
export const APP_SETTINGS_FETCH_SUCCESS = '@appsettings/FETCH_SUCCESS';
export const APP_SETTINGS_FETCH_FAILURE = '@appsettings/FETCH_FAILURE';

export const APP_SETTINGS_UPDATE_REQUEST = '@appsettings/UPDATE_REQUEST';
export const APP_SETTINGS_UPDATE_SUCCESS = '@appsettings/UPDATE_SUCCESS';
export const APP_SETTINGS_UPDATE_FAILURE = '@appsettings/UPDATE_FAILURE';
