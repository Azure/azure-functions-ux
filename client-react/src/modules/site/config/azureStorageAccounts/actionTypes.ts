export const AREA_STRING = '@azureStorageMount';
export const AZURE_STORAGE_MOUNT_FETCH_REQUEST = '@azureStorageMount/FETCH_REQUEST';
export const AZURE_STORAGE_MOUNT_FETCH_SUCCESS = '@azureStorageMount/FETCH_SUCCESS';
export const AZURE_STORAGE_MOUNT_FETCH_FAILURE = '@azureStorageMount/FETCH_FAILURE';

export const AZURE_STORAGE_MOUNT_UPDATE_REQUEST = '@azureStorageMount/UPDATE_REQUEST';
export const AZURE_STORAGE_MOUNT_UPDATE_SUCCESS = '@azureStorageMount/UPDATE_SUCCESS';
export const AZURE_STORAGE_MOUNT_UPDATE_FAILURE = '@azureStorageMount/UPDATE_FAILURE';

export const UPDATE_AZURE_STORAGE_MOUNT_FROM_SITE_UPDATE = '@azureStorageMount/UPDATE_AZURE_STORAGE_MOUNT_FROM_SITE_UPDATE';
