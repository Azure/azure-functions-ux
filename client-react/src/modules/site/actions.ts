import { IAction } from '../../models/action';
import { ISiteState } from './reducer';

export const UPDATE_SITE = 'UPDATE_SITE';
export const updateCurrentSite = (state: Partial<ISiteState>): IAction<Partial<ISiteState>> => ({
  payload: state,
  type: UPDATE_SITE,
});

export const UPDATE_SITE_NO_CACHE = 'UPDATE_SITE';
export const updateCurrentSiteNoCache = (state: Partial<ISiteState>): IAction<Partial<ISiteState>> => ({
  payload: state,
  type: UPDATE_SITE_NO_CACHE,
});

export const UPDATE_RESOURCE_ID = 'UPDATE_RESOURCE_ID';
export const updateResourceIdAtomic = (resourceId: string): IAction<string> => ({
  payload: resourceId,
  type: UPDATE_RESOURCE_ID,
});
