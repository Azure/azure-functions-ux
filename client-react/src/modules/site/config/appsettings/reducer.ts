import { combineReducers } from 'redux';
import { ActionType } from 'typesafe-actions';

import { ArmObj } from '../../../../models/WebAppModels';
import { metadataReducer } from '../../../ApiReducerHelper';
import { ApiState } from '../../../types';
import * as actions from './actions';
import { APP_SETTINGS_FETCH_SUCCESS, APP_SETTINGS_UPDATE_SUCCESS, AREA_STRING, UPDATE_APP_SETTINGS_FROM_SITE_UPDATE } from './actionTypes';

export type AppSettings = { [key: string]: string };

export type AppSettingsActions = ActionType<typeof actions>;
export type AppSettingsState = ApiState<ArmObj<AppSettings>>;

export const InitialState = {
  data: {
    id: '',
    properties: {},
    name: '',
    location: '',
    kind: '',
  },
};

export default combineReducers<AppSettingsState, AppSettingsActions>({
  metadata: metadataReducer(AREA_STRING),
  data: (state = InitialState.data, action) => {
    switch (action.type) {
      case APP_SETTINGS_FETCH_SUCCESS:
        return action.appSettings;
      case APP_SETTINGS_UPDATE_SUCCESS:
        return action.appSettings;
      case UPDATE_APP_SETTINGS_FROM_SITE_UPDATE:
        return { ...state, properties: action.appSettings ? action.appSettings : state.properties };
      default:
        return state;
    }
  },
});
