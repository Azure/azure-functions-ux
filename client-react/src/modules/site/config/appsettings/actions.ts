import { createStandardAction } from 'typesafe-actions';

import { ArmObj, Site } from '../../../../models/WebAppModels';
import {
  APP_SETTINGS_FETCH_FAILURE,
  APP_SETTINGS_FETCH_REQUEST,
  APP_SETTINGS_FETCH_SUCCESS,
  APP_SETTINGS_UPDATE_FAILURE,
  APP_SETTINGS_UPDATE_REQUEST,
  APP_SETTINGS_UPDATE_SUCCESS,
  UPDATE_APP_SETTINGS_FROM_SITE_UPDATE,
} from './actionTypes';
import { AppSettings } from './reducer';

export const fetchAppSettingsRequest = createStandardAction(APP_SETTINGS_FETCH_REQUEST)();

export const fetchAppSettingsSuccess = createStandardAction(APP_SETTINGS_FETCH_SUCCESS).map((appSettings: ArmObj<AppSettings>) => ({
  appSettings,
}));
export const fetchAppSettingsFailure = createStandardAction(APP_SETTINGS_FETCH_FAILURE).map((error: Error) => ({
  error,
}));

export const updateAppSettingsRequest = createStandardAction(APP_SETTINGS_UPDATE_REQUEST).map((appSettings: ArmObj<AppSettings>) => ({
  appSettings,
}));

export const updateAppSettingsSuccess = createStandardAction(APP_SETTINGS_UPDATE_SUCCESS).map((appSettings: ArmObj<AppSettings>) => ({
  appSettings,
}));

export const updateAppSettingsFailure = createStandardAction(APP_SETTINGS_UPDATE_FAILURE).map((error: Error) => ({
  error,
}));

export const updateAppSettingsFromSiteUpdate = createStandardAction(UPDATE_APP_SETTINGS_FROM_SITE_UPDATE).map((site: ArmObj<Site>) => {
  const appSettingsFromSite = !!site.properties && !!site.properties.siteConfig && site.properties.siteConfig.appSettings;
  if (appSettingsFromSite) {
    const updatedAppSettings: AppSettings = {};
    appSettingsFromSite.forEach(appSetting => {
      updatedAppSettings[appSetting.name] = appSetting.value;
    });
    return { appSettings: updatedAppSettings };
  }
  return { appSettings: null };
});
