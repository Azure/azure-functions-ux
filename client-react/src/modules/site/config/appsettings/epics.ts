import { combineEpics, Epic } from 'redux-observable';
import { from, of } from 'rxjs';
import { catchError, filter, map, switchMap } from 'rxjs/operators';
import { isActionOf } from 'typesafe-actions';

import * as Types from '../../../types';
import {
  fetchAppSettingsFailure,
  fetchAppSettingsRequest,
  fetchAppSettingsSuccess,
  updateAppSettingsFailure,
  updateAppSettingsRequest,
  updateAppSettingsSuccess,
} from './actions';
import { AppSettingsActions } from './reducer';

export const fetchAppSettings: Epic<AppSettingsActions, AppSettingsActions, Types.RootState, Types.Services> = (
  action$,
  store,
  { appSettingsApi }
) =>
  action$.pipe(
    filter(isActionOf(fetchAppSettingsRequest)),
    switchMap(() =>
      from(appSettingsApi.fetchAppSettings(store.value)).pipe(
        map(fetchAppSettingsSuccess),
        catchError(err => of(fetchAppSettingsFailure(err)))
      )
    )
  );

export const updateAppSettings: Epic<AppSettingsActions, AppSettingsActions, Types.RootState, Types.Services> = (
  action$,
  store,
  { appSettingsApi }
) =>
  action$.pipe(
    filter(isActionOf(updateAppSettingsRequest)),
    switchMap(action =>
      from(appSettingsApi.updateAppSettings(store.value, action.appSettings)).pipe(
        map(updateAppSettingsSuccess),
        catchError(err => of(updateAppSettingsFailure(err)))
      )
    )
  );
export default combineEpics(fetchAppSettings, updateAppSettings);
