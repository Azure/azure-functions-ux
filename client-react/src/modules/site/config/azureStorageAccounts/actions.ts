import { createStandardAction } from 'typesafe-actions';

import { ArmObj, Site } from '../../../../models/WebAppModels';
import {
  AZURE_STORAGE_MOUNT_FETCH_FAILURE,
  AZURE_STORAGE_MOUNT_FETCH_REQUEST,
  AZURE_STORAGE_MOUNT_FETCH_SUCCESS,
  AZURE_STORAGE_MOUNT_UPDATE_FAILURE,
  AZURE_STORAGE_MOUNT_UPDATE_REQUEST,
  AZURE_STORAGE_MOUNT_UPDATE_SUCCESS,
  UPDATE_AZURE_STORAGE_MOUNT_FROM_SITE_UPDATE,
} from './actionTypes';
import { ArmAzureStorageMount } from './reducer';

export const fetchAzureStorageMountRequest = createStandardAction(AZURE_STORAGE_MOUNT_FETCH_REQUEST)();

export const fetchAzureStorageMountSuccess = createStandardAction(AZURE_STORAGE_MOUNT_FETCH_SUCCESS).map(
  (azureMount: ArmObj<ArmAzureStorageMount>) => ({
    azureMount,
  })
);
export const fetchAzureStorageMountFailure = createStandardAction(AZURE_STORAGE_MOUNT_FETCH_FAILURE).map((error: Error) => ({
  error,
}));

export const updateAzureStorageMountRequest = createStandardAction(AZURE_STORAGE_MOUNT_UPDATE_REQUEST).map(
  (azureMount: ArmObj<ArmAzureStorageMount>) => ({
    azureMount,
  })
);

export const updateAzureStorageMountSuccess = createStandardAction(AZURE_STORAGE_MOUNT_UPDATE_SUCCESS).map(
  (azureMount: ArmObj<ArmAzureStorageMount>) => ({
    azureMount,
  })
);

export const updateAzureStorageMountFailure = createStandardAction(AZURE_STORAGE_MOUNT_UPDATE_FAILURE).map((error: Error) => ({
  error,
}));

export const updateAzureStorageMountFromSiteUpdate = createStandardAction(UPDATE_AZURE_STORAGE_MOUNT_FROM_SITE_UPDATE).map(
  (site: ArmObj<Site>) => {
    const azureStorageMount = !!site.properties && !!site.properties.siteConfig && site.properties.siteConfig.azureStorageAccounts;
    return { azureStorageMount };
  }
);
