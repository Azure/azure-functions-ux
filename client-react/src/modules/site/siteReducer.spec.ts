import mockAxios from 'jest-mock-axios';
import { ActionsObservable } from 'redux-observable';
import { toArray } from 'rxjs/operators';

import rootReducer from '../';
import { IStartupInfo } from '../../models/portal-models';
import { ArmObj, Site } from '../../models/WebAppModels';
import { getStartupInfoAction } from '../portal/actions';
import { RootState, Services } from '../types';
import {
  fetchSiteFailure,
  fetchSiteRequest,
  fetchSiteSuccess,
  updateResourceId,
  updateSiteFailure,
  updateSiteRequest,
  updateSiteSuccess,
} from './actions';
import { SITE_FETCH_FAILURE, SITE_FETCH_SUCCESS, SITE_UPDATE_FAILURE, SITE_UPDATE_SUCCESS } from './actionTypes';
import { fetchSiteFlow, updateSiteFlow } from './epics';
import reducer from './reducer';
import siteApi from './siteApiService';

const testResult = { name: 'fromApi' } as ArmObj<Site>;
describe('Site Store Epics', () => {
  const successDeps = {
    siteApi: {
      fetchSite: async (state: RootState): Promise<ArmObj<Site>> => {
        return testResult;
      },
      updateSite: async (state: RootState, site: ArmObj<Site>): Promise<ArmObj<Site>> => {
        return testResult;
      },
    },
  } as Services;

  const failDeps = {
    siteApi: {
      fetchSite: async (state: RootState): Promise<ArmObj<Site>> => {
        throw new Error('failuremessage');
      },
      updateSite: async (state: RootState, site: ArmObj<Site>): Promise<ArmObj<Site>> => {
        throw new Error('failuremessage');
      },
    },
  } as Services;
  it('Sends Success Action with site object on Successful Fetch', async () => {
    let action$ = ActionsObservable.of(fetchSiteRequest());
    const output$ = fetchSiteFlow(action$, {} as any, successDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(1);
    let action = actions[0];
    if (action.type === SITE_FETCH_SUCCESS) {
      expect(action.site.name).toBe('fromApi');
    }
  });

  it('Sends Success Action with site object on Successful Update', async () => {
    let action$ = ActionsObservable.of(updateSiteRequest(testResult));
    const output$ = updateSiteFlow(action$, {} as any, successDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(4);
    let siteUpdateAction = actions[0];
    if (siteUpdateAction.type === SITE_UPDATE_SUCCESS) {
      expect(siteUpdateAction.site.name).toBe('fromApi');
    }
  });

  it('Sends Error Action with error on failed Fetch', async () => {
    let action$ = ActionsObservable.of(fetchSiteRequest());
    const output$ = fetchSiteFlow(action$, {} as any, failDeps);
    const actions = await output$.pipe(toArray()).toPromise();

    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(SITE_FETCH_FAILURE);
    if (action.type === SITE_FETCH_FAILURE) {
      expect(action.error.message).toBe('failuremessage');
    }
  });

  it('Sends Error Action with error on failed Update', async () => {
    let action$ = ActionsObservable.of(updateSiteRequest(testResult));
    const output$ = updateSiteFlow(action$, {} as any, failDeps);
    const actions = await output$.pipe(toArray()).toPromise();

    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(SITE_UPDATE_FAILURE);
    if (action.type === SITE_UPDATE_FAILURE) {
      expect(action.error.message).toBe('failuremessage');
    }
  });
});

describe('Site Store Reducer', () => {
  const initialState = reducer(undefined, {} as any);
  describe('initial state', () => {
    it('should match a snapshot', () => {
      expect(initialState).toMatchSnapshot();
    });

    it('loading and updating should be false', () => {
      expect(initialState.metadata.loading).toBe(false);
      expect(initialState.metadata.updating).toBe(false);
    });
  });

  describe('Site Fetch Stories', () => {
    it('should trigger loading the fetch is requested', () => {
      const action = fetchSiteRequest();
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(true);
    });

    it('update resourceId action should reflect in state', () => {
      const action = updateResourceId('newresourceid');
      const state = reducer(initialState, action);
      expect(state.resourceId).toBe('newresourceid');
    });

    it('metadata and site should update on successful load', () => {
      const action = fetchSiteSuccess({ name: 'testsite' } as any);
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(false);
      expect(state.data.name).toBe('testsite');
    });

    it('error should be reflected on failed load', () => {
      const action = fetchSiteFailure(new Error('testerror'));
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(false);
      expect(state.metadata.fetchError).toBe(true);
      expect(state.metadata.fetchErrorObject.message).toBe('testerror');
    });
  });

  describe('Site Fetch Stories', () => {
    it('should trigger updating the fetch is requested', () => {
      const action = updateSiteRequest(testResult);
      const state = reducer(initialState, action);
      expect(state.metadata.updating).toBe(true);
    });

    it('metadata and site should update on successful update', () => {
      const action = updateSiteSuccess({ name: 'testsite' } as any);
      const state = reducer(initialState, action);
      expect(state.metadata.updating).toBe(false);
      expect(state.data.name).toBe('testsite');
    });

    it('error should be reflected on failed load', () => {
      const action = updateSiteFailure(new Error('testerror'));
      const state = reducer(initialState, action);
      expect(state.metadata.updating).toBe(false);
      expect(state.metadata.updateError).toBe(true);
      expect(state.metadata.updateErrorObject.message).toBe('testerror');
    });
  });

  describe('Site Service', () => {
    const initialState = rootReducer(undefined, {} as any);
    const catchFn = jest.fn();
    const thenFn = jest.fn();
    let state;
    beforeEach(() => {
      const updateResourceIdAction = updateResourceId('resourceid');
      const updateSUIAction = getStartupInfoAction({
        token: 'testtoken',
        armEndpoint: 'testEndpoint',
      } as IStartupInfo);

      state = rootReducer(rootReducer(initialState, updateResourceIdAction), updateSUIAction);
    });

    afterEach(() => {
      mockAxios.reset();
      thenFn.mockClear();
      catchFn.mockClear();
    });

    it('Fetch Api calls api with appropriate info', async () => {
      const fetcher = siteApi.fetchSite(state);
      let siteApiRequestInfo = mockAxios.lastReqGet();
      expect(mockAxios).toHaveBeenCalledWith({
        method: 'GET',
        url: 'testEndpointresourceid?api-version=2018-02-01',
        data: null,
        headers: {
          Authorization: `Bearer testtoken`,
        },
      });
      let responseObj = { data: { name: 'testsite' } };
      mockAxios.mockResponse(responseObj, siteApiRequestInfo);
      const result = await fetcher;
      expect(result.name).toBe('testsite');
    });

    it('Update Api calls api with appropriate info', async () => {
      const fetcher = siteApi.updateSite(state, testResult);
      let siteApiRequestInfo = mockAxios.lastReqGet();
      expect(mockAxios).toHaveBeenCalledWith({
        method: 'PUT',
        url: 'testEndpointresourceid?api-version=2018-02-01',
        data: testResult,
        headers: {
          Authorization: `Bearer testtoken`,
        },
      });
      let responseObj = { data: { name: 'testsite' } };
      mockAxios.mockResponse(responseObj, siteApiRequestInfo);
      const result = await fetcher;
      expect(result.name).toBe('testsite');
    });

    it('Fetch Api should throw on error', async () => {
      const fetcher = siteApi
        .fetchSite(state)
        .then(thenFn)
        .catch(catchFn);
      mockAxios.mockError(new Error('errorMessage'));
      await fetcher;
      expect(thenFn).not.toHaveBeenCalled();
      expect(catchFn).toHaveBeenCalled();
    });

    it('Update Api should throw on error', async () => {
      const fetcher = siteApi
        .updateSite(state, testResult)
        .then(thenFn)
        .catch(catchFn);
      mockAxios.mockError(new Error('errorMessage'));
      await fetcher;
      expect(thenFn).not.toHaveBeenCalled();
      expect(catchFn).toHaveBeenCalled();
    });
  });
});
