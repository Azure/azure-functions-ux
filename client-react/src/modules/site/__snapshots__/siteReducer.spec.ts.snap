// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site Store Reducer initial state should match a snapshot 1`] = `
Object {
  "data": Object {
    "id": "",
    "kind": "",
    "location": "",
    "name": "",
    "properties": Object {},
  },
  "metadata": Object {
    "fetchError": false,
    "fetchErrorObject": [Error],
    "loading": false,
    "updateError": false,
    "updateErrorObject": [Error],
    "updating": false,
  },
  "resourceId": "",
}
`;
